modules/core/core-content/src/main/java/com/enonic/xp/core/impl/content/ParentContentSynchronizer.java

```diff
@@ -46,220 +50,269 @@ public ParentContentSynchronizer( @Reference final ContentService contentService
 
-        sourceContext.runWith( () -> {
+        final Map<NodePath, Context> targetContexts = !sourceContents.isEmpty() ? initContexts(
+            ProjectName.from( sourceContents.stream().findAny().get().getTargetContext().getRepositoryId() ) ) : Map.of();
 
-            queue.add( sourceContent );
+        final List<ContentToSync> contentsToSync = sourceContents.stream().filter( sourceContent -> {
 
-            final Content root = contentService.getByPath( ContentPath.ROOT );
+            final Content root = sourceContent.getSourceContext().callWith( () -> contentService.getByPath( ContentPath.ROOT ) );
+            return !root.getId().equals( sourceContent.getSourceContent().getId() );
 
-            if ( !root.getId().equals( sourceContent.getId() ) )
-            {
-                this.doSync( sourceContent, sourceContext, targetContext );
-            }
+        } ).collect( Collectors.toList() );
 
-            while ( queue.size() > 0 )
-            {
-                final Content currentContent = queue.poll();
+        if ( !contentsToSync.isEmpty() )
+        {
+            this.doSync( contentsToSync );
+        }
 
-                final FindContentByParentResult result = contentService.findByParent( FindContentByParentParams.create().
-                    parentId( currentContent.getId() ).
-                    recursive( false ).
-                    childOrder( currentContent.getChildOrder() ).
-                    size( -1 ).
-                    build() );
+        while ( queue.size() > 0 )
+        {
+            final ContentToSync currentContentToSync = queue.poll();
 
-                for ( final Content content : result.getContents() )
-                {
-                    this.doSync( content, sourceContext, targetContext );
+            final FindContentByParentResult result = currentContentToSync.getSourceContext()
+                .callWith( () -> contentService.findByParent( FindContentByParentParams.create()
+                                                                  .parentId( currentContentToSync.getId() )
+                                                                  .recursive( false )
+                                                                  .childOrder( currentContentToSync.getSourceContent().getChildOrder() )
+                                                                  .size( -1 )
+                                                                  .build() ) );
 
-                    if ( content.hasChildren() )
+            if ( result.getContents().isNotEmpty() )
+            {
+                final List<ContentToSync> childrenToSync = result.getContents().stream().map( content -> {
+                    final Context actualTargetContext = getActualContext( content.getId(), targetContexts.values() );
+
+                    return ContentToSync.create()
+                        .sourceContent( content )
+                        .targetContent( actualTargetContext != null
```

Comment:
when actualTargetContent is null?

Reply:
It's null if we sync new content in a project (source).

Reply:
tested